File: Simulator.m
classdef Simulator < handle
    % Simulator 仿真器
    
    properties
        MEC             % MEC实例
        TaskManager     % 任务管理器
        LyapunovManager % 李雅普诺夫队列管理器
        Scheduler       % 调度器
        CurrentTimeSlot % 当前时隙
        TotalTimeSlots  % 总时隙数
        Statistics      % 仿真统计
    end
    
    methods
        function obj = Simulator(totalTimeSlots)
            % 构造函数
            if nargin < 1
                totalTimeSlots = 1000;
            end
            
            obj.MEC = MEC();
            obj.TaskManager = TaskManager();
            obj.LyapunovManager = LyapunovManager();
            obj.Scheduler = Scheduler(constants.GreedySchedule, constants.VV_DEFAULT); % 默认使用贪心调度
            obj.CurrentTimeSlot = 0;
            obj.TotalTimeSlots = totalTimeSlots;
            obj.Statistics = SimulationStats();
            
            % 初始化任务类型统计
            K = constants.K();
            for i = 1:K
                obj.Statistics.TaskTypeStats(i) = TaskTypeStat();
            end
        end
        
        function setCacheStrategy(obj, strategy)
            % 设置缓存策略
            obj.MEC.setCacheStrategy(strategy);
        end
        
        function setScheduleStrategy(obj, algorithm, vv)
            % 设置调度策略 和 李雅普诺夫漂移参数
            if nargin < 3
                vv = constants.VV_DEFAULT;
            end
            obj.Scheduler = Scheduler(algorithm, vv);
        end
        
        function runSimulation(obj)
            % 运行仿真
            fprintf('开始仿真，总时隙数: %d\n', obj.TotalTimeSlots);
            
            for t = 0:(obj.TotalTimeSlots-1)
                obj.CurrentTimeSlot = t;
                obj.runTimeSlot();
                
                % 每100个时隙输出一次进度
                if mod(t+1, 100) == 0
                    fprintf('时隙进度: %d/%d\n', t+1, obj.TotalTimeSlots);
                end
            end
            
            obj.printStatistics();
        end
        
        function runTimeSlot(obj)
            % 运行单个时隙
            % 更新MEC的当前时隙
            obj.MEC.updateTimeSlot(obj.CurrentTimeSlot);
            
            % 用于当前时隙收益计算的数据
            scheduledTasks = containers.Map('KeyType', 'int32', 'ValueType', 'any');
            completedTasks = containers.Map('KeyType', 'int32', 'ValueType', 'any');
            cacheHitTasks = containers.Map('KeyType', 'int32', 'ValueType', 'any');

            % 1. 生成新任务并处理过期任务
            newTasks = obj.TaskManager.generateRandomTasks(obj.CurrentTimeSlot);
            obj.Statistics.TotalTasksGenerated = obj.Statistics.TotalTasksGenerated + length(newTasks);
            
            % 更新任务类型生成统计和访问统计
            for i = 1:length(newTasks)
                task = newTasks{i};
                stat = obj.Statistics.TaskTypeStats(task.TaskType);
                stat.Generated = stat.Generated + 1;
                % 记录任务访问（正确的访问统计方式）
                obj.MEC.recordTaskAccess(task.TaskType);
            end
            
            % 移除过期任务
            expiredCounts = obj.TaskManager.removeExpiredTasks(obj.CurrentTimeSlot);
            keys = cell2mat(expiredCounts.keys);
            for i = 1:length(keys)
                taskType = keys(i);
                count = expiredCounts(taskType);
                obj.Statistics.TotalTasksDropped = obj.Statistics.TotalTasksDropped + count;
                stat = obj.Statistics.TaskTypeStats(taskType);
                stat.Dropped = stat.Dropped + count;
            end
            
            % 2. 统一处理：先把所有新生成的任务都放到积压队列中
            for i = 1:length(newTasks)
                task = newTasks{i};
                obj.TaskManager.addToBacklog(task);
                obj.Statistics.TotalCacheAccess = obj.Statistics.TotalCacheAccess + 1;
            end
            
            % 3. 时隙开始检查：如果任务类型缓存命中，清空该类型积压队列
            K = constants.K();
            for taskType = 1:K
                backlogCount = obj.TaskManager.getBacklogCount(taskType);
                if backlogCount > 0
                    if obj.MEC.isCacheHit(taskType)
                        % 缓存命中，该类型所有积压任务直接完成
                        obj.Statistics.CacheHitCount = obj.Statistics.CacheHitCount + backlogCount;
                        obj.Statistics.TotalTasksCompleted = obj.Statistics.TotalTasksCompleted + backlogCount;
                        stat = obj.Statistics.TaskTypeStats(taskType);
                        stat.CacheHits = stat.CacheHits + backlogCount;
                        stat.Completed = stat.Completed + backlogCount;
                        
                        % 记录缓存命中任务的总优先级
                        if obj.TaskManager.TaskTypes.isKey(taskType)
                            tt = obj.TaskManager.TaskTypes(taskType);
                            stat.CacheHitPrioritySum = stat.CacheHitPrioritySum + (backlogCount * tt.Priority);
                            
                            % 记录用于收益计算
                            hitInfo.backlogCount = backlogCount;
                            hitInfo.metaK = tt.MetaK;
                            cacheHitTasks(taskType) = hitInfo;
                        end
                        
                        obj.TaskManager.removeTasksFromBacklog(taskType, backlogCount);
                    end
                end
            end
            
            % 4. 调度积压队列中的任务
            schedulingResults = obj.Scheduler.scheduleTasks(obj.MEC, obj.TaskManager, obj.LyapunovManager);
            for i = 1:length(schedulingResults)
                res = schedulingResults{i};
                % 记录用于收益计算
                if obj.TaskManager.TaskTypes.isKey(res.TaskType) && res.NodeID > 0
                    tt = obj.TaskManager.TaskTypes(res.TaskType);
                    taskInfo.node = obj.MEC.VirtualNodes{res.NodeID};
                    taskInfo.mkr = res.MKR;
                    taskInfo.ck = tt.Ck;
                    taskInfo.backlogCount = res.CompletedTasks; % 假设调度一个任务就完成一个
                    scheduledTasks(res.TaskType) = taskInfo;
                end
            end
                  
            % 5. 更新虚拟节点状态
            completedTaskTypesMap = obj.MEC.updateNodes();
            completedTaskTypes = cell2mat(keys(completedTaskTypesMap));

            % 遍历所有类型积压队列，清空completedTaskTypes中存在的任务类型对应的积压队列
            % TotalTasksCompleted的统计应该等于对应的积压队列中的任务
            for i = 1:length(completedTaskTypes)
                taskType = completedTaskTypes(i);
                backlogCount = obj.TaskManager.getBacklogCount(taskType);
                
                if backlogCount > 0
                    obj.Statistics.TotalTasksCompleted = obj.Statistics.TotalTasksCompleted + backlogCount;
                    stat = obj.Statistics.TaskTypeStats(taskType);
                    stat.Completed = stat.Completed + backlogCount;
                    
                    % 记录用于收益计算
                    completedTasks(taskType) = backlogCount;

                    obj.TaskManager.removeTasksFromBacklog(taskType, backlogCount);
                end
            end
            
            % 更新任务类型完成统计 (这部分逻辑在上面处理积压队列时已完成)
            
            % 6. 缓存更新：完成的任务类型添加到缓存
            for i = 1:length(completedTaskTypes)
                taskType = completedTaskTypes(i);
                if obj.TaskManager.TaskTypes.isKey(taskType)
                    tt = obj.TaskManager.TaskTypes(taskType);
                    obj.MEC.addToCache(taskType, tt.MetaK, obj.TaskManager);
                end
            end
            
            % 7. 更新李雅普诺夫队列
            for taskType = 1:K
                % 计算bk(t) - 本时隙完成的该类型任务数
                bk = 0;
                for j = 1:length(schedulingResults)
                    if schedulingResults{j}.TaskType == taskType
                        bk = bk + schedulingResults{j}.CompletedTasks;
                    end
                end
                
                % 计算ak(t) - 本时隙新到达的该类型任务数
                ak = 0;
                for j = 1:length(newTasks)
                    if newTasks{j}.TaskType == taskType
                        ak = ak + 1;
                    end
                end
                
                % 计算丢弃的任务数
                droppedCount = 0;
                if expiredCounts.isKey(taskType)
                    droppedCount = expiredCounts(taskType);
                end
                
                obj.LyapunovManager.updateQueue(taskType, bk, droppedCount, ak, obj.TaskManager);
            end
            
            % 8. 更新收益
            obj.MEC.updateRevenue(obj.TaskManager, scheduledTasks, completedTasks, cacheHitTasks);
            obj.Statistics.TotalRevenue = obj.MEC.Revenue; % 直接使用MEC累计的总收益
            
            % 记录用于绘图的时序数据
            obj.Statistics.recordTimeseriesData(obj.CurrentTimeSlot + 1, obj.MEC, obj.TaskManager, obj.LyapunovManager);
        end
        
        function printStatistics(obj)
            % 打印统计信息
            fprintf('\n=== 仿真统计结果 ===\n');
            fprintf('总时隙数: %d\n', obj.TotalTimeSlots);
            fprintf('总生成任务数: %d\n', obj.Statistics.TotalTasksGenerated);
            fprintf('总完成任务数: %d\n', obj.Statistics.TotalTasksCompleted);
            fprintf('总丢弃任务数: %d\n', obj.Statistics.TotalTasksDropped);
            fprintf('任务完成率: %.2f%%\n', obj.Statistics.TotalTasksCompleted / obj.Statistics.TotalTasksGenerated * 100);
            fprintf('缓存命中次数: %d\n', obj.Statistics.CacheHitCount);
            fprintf('总缓存访问次数: %d\n', obj.Statistics.TotalCacheAccess);
            
            if obj.Statistics.TotalCacheAccess > 0
                fprintf('缓存命中率: %.2f%%\n', obj.Statistics.CacheHitCount / obj.Statistics.TotalCacheAccess * 100);
            else
                fprintf('缓存命中率: 0.00%%\n');
            end
            
            fprintf('缓存利用率: %.2f%%\n', obj.MEC.getCacheUtilization() * 100);
            fprintf('节点利用率: %.2f%%\n', obj.MEC.getNodeUtilization() * 100);
            fprintf('最终收益: %.2f\n', obj.MEC.Revenue);
            fprintf('最终收入: %.2f\n', obj.MEC.Income);
            fprintf('最终成本: %.2f\n', obj.MEC.Cost);
            
            % 计算平均队列长度
            totalLyapunovQueueLength = 0;
            K = constants.K();
            for i = 1:K
                totalLyapunovQueueLength = totalLyapunovQueueLength + obj.LyapunovManager.getQueueLength(i);
            end
            obj.Statistics.AverageLyapunovQueueLength = totalLyapunovQueueLength / K;
            fprintf('平均队列长度: %.2f\n', obj.Statistics.AverageLyapunovQueueLength);
            % 挤压队列平均长度
            totalBacklogQueueLength = 0;
            for i = 1:K
                totalBacklogQueueLength = totalBacklogQueueLength + obj.TaskManager.getBacklogCount(i);
            end
            obj.Statistics.AverageBacklogQueueLength = totalBacklogQueueLength / K;
            fprintf('平均挤压队列长度: %.2f\n', obj.Statistics.AverageBacklogQueueLength);
            
            % 打印各任务类型统计
            fprintf('\n=== 任务类型统计 ===\n');
            keys = cell2mat(obj.Statistics.TaskTypeStats.keys);
            for i = 1:length(keys)
                taskType = keys(i);
                stat = obj.Statistics.TaskTypeStats(taskType);
                fprintf('任务类型 %d: 生成=%d, 完成=%d, 丢弃=%d, 缓存命中=%d\n', ...
                    taskType, stat.Generated, stat.Completed, stat.Dropped, stat.CacheHits);
            end
        end
        
        function stats = getStatistics(obj)
            % 获取统计信息
            stats = obj.Statistics;
        end
    end
end

File: LYAPUNOV/simulator.py
"""
仿真器类
"""

# 处理导入问题
try:
    from .constants import Constants
    from .mec import MEC
    from .task_manager import TaskManager
    from .lyapunov_classes import LyapunovManager
    from .scheduler import Scheduler
    from .stats_classes import SimulationStats
except ImportError:
    from constants import Constants
    from mec import MEC
    from task_manager import TaskManager
    from lyapunov_classes import LyapunovManager
    from scheduler import Scheduler
    from stats_classes import SimulationStats

# 导入日志工具
try:
     from .logger import logger
except ImportError:
     from logger import logger
       

class Simulator:
    """Simulator 仿真器"""
    
    def __init__(self, total_time_slots=1000):
        """构造函数"""
        self.MEC = MEC()
        self.TaskManager = TaskManager()
        self.LyapunovManager = LyapunovManager()
        self.Scheduler = Scheduler(Constants.GreedySchedule, Constants.VV_DEFAULT)  # 默认使用贪心调度
        self.CurrentTimeSlot = 0
        self.TotalTimeSlots = total_time_slots
        self.Statistics = SimulationStats()
        
    def set_cache_strategy(self, strategy):
        """设置缓存策略"""
        self.MEC.set_cache_strategy(strategy)
        
    def set_schedule_strategy(self, algorithm, vv=None):
        """设置调度策略 和 李雅普诺夫漂移参数"""
        if vv is None:
            vv = Constants.VV_DEFAULT
        self.Scheduler = Scheduler(algorithm, vv)
        
    def run_simulation(self):
        """运行仿真"""
        print(f'开始仿真，总时隙数: {self.TotalTimeSlots}')
        
        for t in range(self.TotalTimeSlots):
            self.CurrentTimeSlot = t
            self.run_time_slot()
            
            # 每100个时隙输出一次进度
            if (t + 1) % 100 == 0:
                print(f'时隙进度: {t + 1}/{self.TotalTimeSlots}')
        
        self.print_statistics()
        
    def run_time_slot(self):
        """运行单个时隙"""
     
        # 记录时隙开始
        logger.separator("=", 80)
        logger.info(f"开始执行时隙 {self.CurrentTimeSlot}")
        
        # 更新MEC和Scheduler的当前时隙
        self.MEC.update_time_slot(self.CurrentTimeSlot)
        self.Scheduler.update_time_slot(self.CurrentTimeSlot)
        
        # 用于当前时隙收益计算的数据
        scheduled_tasks = {}
        completed_tasks = {}
        cache_hit_tasks = {}

        # 1. 生成新任务并处理过期任务
        new_tasks = self.TaskManager.generate_random_tasks(self.CurrentTimeSlot)
        self.Statistics.TotalTasksGenerated += len(new_tasks)
        logger.info(f"生成新任务数量: {len(new_tasks)}")
        
        # 更新任务类型生成统计和访问统计
        for task in new_tasks:
            stat = self.Statistics.TaskTypeStats[task.TaskType]
            stat.Generated += 1
            # 记录任务访问（正确的访问统计方式）
            self.MEC.record_task_access(task.TaskType)
        
        # 移除过期任务
        expired_counts = self.TaskManager.remove_expired_tasks(self.CurrentTimeSlot)
        total_expired = sum(expired_counts.values())
        if total_expired > 0:
            logger.info(f"移除过期任务数量: {total_expired}")
        
        for task_type, count in expired_counts.items():
            self.Statistics.TotalTasksDropped += count
            stat = self.Statistics.TaskTypeStats[task_type]
            stat.Dropped += count
        
        # 2. 统一处理：先把所有新生成的任务都放到积压队列中
        for task in new_tasks:
            self.TaskManager.add_to_backlog(task)
            self.Statistics.TotalCacheAccess += 1
        
        # 3. 时隙开始检查：如果任务类型缓存命中，清空该类型积压队列
        K = Constants.K()
        for task_type in range(1, K + 1):
            backlog_count = self.TaskManager.get_backlog_count(task_type)
            if backlog_count > 0:
                if self.MEC.is_cache_hit(task_type):
                    # 缓存命中，该类型所有积压任务直接完成
                    self.Statistics.CacheHitCount += backlog_count
                    self.Statistics.TotalTasksCompleted += backlog_count
                    stat = self.Statistics.TaskTypeStats[task_type]
                    stat.CacheHits += backlog_count
                    stat.Completed += backlog_count
                    
                    # 记录缓存命中任务的总优先级
                    if task_type in self.TaskManager.TaskTypes:
                        tt = self.TaskManager.TaskTypes[task_type]
                        stat.CacheHitPrioritySum += (backlog_count * tt.Priority)
                        
                        # 记录用于收益计算
                        hit_info = {
                            'backlogCount': backlog_count,
                            'metaK': tt.MetaK
                        }
                        cache_hit_tasks[task_type] = hit_info
                    
                    self.TaskManager.remove_tasks_from_backlog(task_type, backlog_count)
        
        # 4. 调度积压队列中的任务
        scheduling_results = self.Scheduler.schedule_tasks(self.MEC, self.TaskManager, self.LyapunovManager)
        if len(scheduling_results) > 0:
            logger.info(f"调度任务数量: {len(scheduling_results)}")
        
        for res in scheduling_results:
            # 记录用于收益计算
            if res.TaskType in self.TaskManager.TaskTypes and res.NodeID > 0:
                tt = self.TaskManager.TaskTypes[res.TaskType]
                task_info = {
                    'node': self.MEC.VirtualNodes[res.NodeID - 1],  # 转换为0基索引
                    'mkr': res.MKR,
                    'ck': tt.Ck,
                    'backlogCount': res.CompletedTasks,  # 假设调度一个任务就完成一个
                    'bkr': res.Bkr
                }
                scheduled_tasks[res.TaskType] = task_info
                  
        # 5. 更新虚拟节点状态
        completed_task_types_map = self.MEC.update_nodes()
        total_completed_types = len(completed_task_types_map)
        if total_completed_types > 0:
            logger.info(f"完成计算的任务类型数: {total_completed_types}")

        # 遍历所有类型积压队列，清空completedTaskTypes中存在的任务类型对应的积压队列
        # TotalTasksCompleted的统计应该等于对应的积压队列中的任务
        for task_type, _ in completed_task_types_map.items():
            backlog_count = self.TaskManager.get_backlog_count(task_type)
            
            if backlog_count > 0:
                self.Statistics.TotalTasksCompleted += backlog_count
                stat = self.Statistics.TaskTypeStats[task_type]
                stat.Completed += backlog_count
                
                # 记录用于收益计算
                completed_tasks[task_type] = backlog_count

                self.TaskManager.remove_tasks_from_backlog(task_type, backlog_count)
             
        # 6. 缓存更新：完成的任务类型添加到缓存
        for task_type in completed_task_types_map:
            if task_type in self.TaskManager.TaskTypes:
                tt = self.TaskManager.TaskTypes[task_type]
                self.MEC.add_to_cache(task_type, tt.MetaK, self.TaskManager)
        
        # 7. 更新李雅普诺夫队列
        for task_type in range(1, K + 1):
            # 计算bk(t) - 本时隙完成的该类型任务数
            bk = 0
            scheduled_mkr = None  # 实际调度的任务的mkr值
            for res in scheduling_results:
                if res.TaskType == task_type:
                    bk += res.Bkr
                    if scheduled_mkr is None:  # 只取第一个调度结果的mkr值
                        scheduled_mkr = res.MKR
            
            # 计算ak(t) - 本时隙新到达的该类型任务数
            ak = 0
            for task in new_tasks:
                if task.TaskType == task_type:
                    ak += 1
            
            # 计算丢弃的任务数
            dropped_count = 0
            if task_type in expired_counts:
                dropped_count = expired_counts[task_type]
            
            # 传递实际调度的任务的mkr值和当前时隙
            self.LyapunovManager.update_queue(task_type, bk, dropped_count, ak, self.TaskManager, scheduled_mkr, self.CurrentTimeSlot)
        
        # 8. 更新收益
        self.MEC.update_revenue(self.TaskManager, scheduled_tasks, completed_tasks, cache_hit_tasks)
        self.Statistics.TotalRevenue = self.MEC.Revenue  # 直接使用MEC累计的总收益
        
        # 更新平均收益
        if self.CurrentTimeSlot > 0:
            self.Statistics.AverageRevenue = self.Statistics.TotalRevenue / (self.CurrentTimeSlot + 1)
        else:
            self.Statistics.AverageRevenue = self.Statistics.TotalRevenue
        
        # 记录用于绘图的时序数据
        self.Statistics.record_timeseries_data(self.CurrentTimeSlot + 1, self.MEC, self.TaskManager)
        
        # 更新积压队列长度统计
        self.Statistics.update_backlog_stats(self.TaskManager)
        
        # 记录时隙结束信息
        total_backlog = self.TaskManager.get_all_backlog_count()
        logger.info(f"时隙 {self.CurrentTimeSlot} 执行完成")
        logger.info(f"当前总积压任务数: {total_backlog}")
        logger.info(f"累计总收益: {self.MEC.Revenue:.6f}")
        logger.separator("=", 80)
        
    def print_statistics(self):
        """打印统计信息"""
        print('\n=== 仿真统计结果 ===')
        print(f'总时隙数: {self.TotalTimeSlots}')
        print(f'总生成任务数: {self.Statistics.TotalTasksGenerated}')
        print(f'总完成任务数: {self.Statistics.TotalTasksCompleted}')
        print(f'总丢弃任务数: {self.Statistics.TotalTasksDropped}')
        
        if self.Statistics.TotalTasksGenerated > 0:
            completion_rate = self.Statistics.TotalTasksCompleted / self.Statistics.TotalTasksGenerated * 100
            print(f'任务完成率: {completion_rate:.2f}%')
        else:
            print('任务完成率: 0.00%')
        
        print(f'缓存命中次数: {self.Statistics.CacheHitCount}')
        print(f'总缓存访问次数: {self.Statistics.TotalCacheAccess}')
        
        if self.Statistics.TotalCacheAccess > 0:
            cache_hit_rate = self.Statistics.CacheHitCount / self.Statistics.TotalCacheAccess * 100
            print(f'缓存命中率: {cache_hit_rate:.2f}%')
        else:
            print('缓存命中率: 0.00%')
        
        print(f'缓存利用率: {self.MEC.get_cache_utilization() * 100:.2f}%')
        print(f'节点利用率: {self.MEC.get_node_utilization() * 100:.2f}%')
        print(f'最终收益: {self.MEC.Revenue:.2f}')
        print(f'最终收入: {self.MEC.Income:.2f}')
        print(f'最终成本: {self.MEC.Cost:.2f}')
        
        print(f'平均队列长度: {self.LyapunovManager.get_all_queue_lengths()}')
        print(f'平均积压队列长度: {self.TaskManager.get_all_backlog_count()}')
        
        # 打印各任务类型统计
        print('\n=== 任务类型统计 ===')
        for task_type in self.Statistics.TaskTypeStats:
            stat = self.Statistics.TaskTypeStats[task_type]
            print(f'任务类型 {task_type}: 生成={stat.Generated}, 完成={stat.Completed}, '
                  f'丢弃={stat.Dropped}, 缓存命中={stat.CacheHits}')
        
    def get_statistics(self):
        """获取统计信息"""
        return self.Statistics
